src/gui/qgsmaptoolcapture.cpp

```diff
@@ -1073,3 +1206,187 @@ void QgsMapToolCapture::updateExtraSnapLayer()
+    QgsGeometry g( std::make_unique<QgsPoint>( savePoint ) );
+
+    // The snapping result needs to be added so it's available in the @snapping_results variable of default value etc. expression contexts
+    addVertex( e->mapPoint(), e->mapPointMatch() );
+
+    geometryCaptured( g );
+    pointCaptured( savePoint );
+
+    stopCapturing();
+
+    // we are done with digitizing for now so instruct advanced digitizing dock to reset its CAD points
+    cadDockWidget()->clearPoints();
+  }
+
+  // LINE AND POLYGON CAPTURING
+  else if ( mode() == CaptureLine || mode() == CapturePolygon )
+  {
+    bool digitizingFinished = false;
+
+    if ( mCurrentCaptureTechnique == Shape )
+    {
+      if ( !mCurrentShapeMapTool )
+      {
+        emit messageEmitted( tr( "Cannot capture a shape without a shape tool defined" ), Qgis::MessageLevel::Warning );
+        return;
+      }
+      else
+      {
+        if ( !mTempRubberBand )
+        {
+          mTempRubberBand.reset( createCurveRubberBand() );
+          mTempRubberBand->setStringType( mLineDigitizingType );
+          mTempRubberBand->setRubberBandGeometryType( mCaptureMode == CapturePolygon ? QgsWkbTypes::PolygonGeometry : QgsWkbTypes::LineGeometry );
+        }
+
+        digitizingFinished = mCurrentShapeMapTool->cadCanvasReleaseEvent( e, mCaptureMode );
+        if ( digitizingFinished )
+          mCurrentShapeMapTool->clean();
+      }
+    }
+    else // i.e. not shape
+    {
+      //add point to list and to rubber band
+      if ( e->button() == Qt::LeftButton )
+      {
+        const int error = addVertex( e->mapPoint(), e->mapPointMatch() );
+        if ( error == 2 )
+        {
+          //problem with coordinate transformation
+          emit messageEmitted( tr( "Cannot transform the point to the layers coordinate system" ), Qgis::MessageLevel::Warning );
+          return;
+        }
+
+        startCapturing();
+      }
+      else if ( e->button() == Qt::RightButton )
+      {
+        // End of string
+        deleteTempRubberBand();
+
+        //lines: bail out if there are not at least two vertices
+        if ( mode() == CaptureLine && size() < 2 )
+        {
+          stopCapturing();
+          return;
+        }
+
+        //polygons: bail out if there are not at least two vertices
+        if ( mode() == CapturePolygon && size() < 3 )
+        {
+          stopCapturing();
+          return;
+        }
+
+        if ( mode() == CapturePolygon || e->modifiers() == Qt::ShiftModifier )
+        {
+          closePolygon();
+        }
+
+        digitizingFinished = true;
+      }
+    }
+
+    if ( digitizingFinished )
+    {
+      QgsGeometry g;
+
+      //does compoundcurve contain circular strings?
+      //does provider support circular strings?
+      const bool hasCurvedSegments = captureCurve()->hasCurvedSegments();
+      const bool providerSupportsCurvedSegments = vlayer && ( vlayer->dataProvider()->capabilities() & QgsVectorDataProvider::CircularGeometries );
```

Comment:
This logic shouldn't be in the base class, but rather in the QgsMapToolCaptureLayerGeometry class.